File: Android/agc-authservice-demo-java/app/src/main/java/com/huawei/agc/quickstart/auth/SettingsActivity.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.agc.quickstart.auth;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import com.huawei.agconnect.auth.AGConnectAuth;
import com.huawei.agconnect.auth.AGConnectAuthCredential;
import com.huawei.agconnect.auth.EmailAuthProvider;
import com.huawei.agconnect.auth.PhoneAuthProvider;
import com.huawei.agconnect.auth.VerifyCodeResult;
import com.huawei.agconnect.auth.VerifyCodeSettings;
import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hmf.tasks.Task;
import com.huawei.hmf.tasks.TaskExecutors;

public class SettingsActivity extends AppCompatActivity {
    private static final String TAG = SettingsActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.settings_activity);
        initView();
    }

    private void initView() {
        findViewById(R.id.layout_delete).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AGConnectAuth.getInstance().deleteUser();
                startActivity(new Intent(SettingsActivity.this, LoginActivity.class));
                SettingsActivity.this.finish();
            }
        });

        findViewById(R.id.layout_update_email).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateEmail();
            }
        });

        findViewById(R.id.layout_update_phone).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updatePhone();
            }
        });

        findViewById(R.id.layout_update_password).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AGConnectAuth.getInstance().getCurrentUser().updatePassword("new password", "verifyCode", AGConnectAuthCredential.Email_Provider);
                //or
                //AGConnectAuth.getInstance().getCurrentUser().updatePassword("new password", "verifyCode", AGConnectAuthCredential.Phone_Provider);
            }
        });

        findViewById(R.id.layout_reset_password).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AGConnectAuth.getInstance().resetPassword("email", "new password", "verifycode");
                //or
                //AGConnectAuth.getInstance().resetPassword("countryCode", "phoneNumber", "new password", "verifycode");
            }
        });
    }

    private void updatePhone() {
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_phone, null, false);
        final AlertDialog dialog = new AlertDialog.Builder(this).setView(view).create();
        EditText editCc = view.findViewById(R.id.et_country_code);
        EditText editAccount = view.findViewById(R.id.et_account);
        EditText editCode = view.findViewById(R.id.et_verify_code);
        Button send = view.findViewById(R.id.btn_send);
        send.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String cc = editCc.getText().toString();
                String phone = editAccount.getText().toString();
                VerifyCodeSettings settings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .build();
                Task<VerifyCodeResult> task = PhoneAuthProvider.requestVerifyCode(cc, phone, settings);
                task.addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<VerifyCodeResult>() {
                    @Override
                    public void onSuccess(VerifyCodeResult verifyCodeResult) {
                    }
                });
            }
        });
        Button update = view.findViewById(R.id.update);
        update.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String cc = editCc.getText().toString();
                String phone = editAccount.getText().toString();
                String verifyCode = editCode.getText().toString();
                AGConnectAuth.getInstance().getCurrentUser().updatePhone(cc, phone, verifyCode)
                    .addOnSuccessListener(new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void aVoid) {
                            Log.d(TAG, "updatePhone success");
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        Log.d(TAG, "updatePhone fail" + e);
                    }
                });
                dialog.dismiss();
            }
        });
        dialog.show();
    }

    private void updateEmail() {
        View view = LayoutInflater.from(this).inflate(R.layout.dialog_email, null, false);
        final AlertDialog dialog = new AlertDialog.Builder(this).setView(view).create();
        EditText editAccount = view.findViewById(R.id.et_account);
        EditText editCode = view.findViewById(R.id.et_verify_code);
        Button send = view.findViewById(R.id.btn_send);
        send.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String email = editAccount.getText().toString();
                VerifyCodeSettings settings = VerifyCodeSettings.newBuilder()
                    .action(VerifyCodeSettings.ACTION_REGISTER_LOGIN)
                    .build();
                Task<VerifyCodeResult> task = EmailAuthProvider.requestVerifyCode(email, settings);
                task.addOnSuccessListener(TaskExecutors.uiThread(), new OnSuccessListener<VerifyCodeResult>() {
                    @Override
                    public void onSuccess(VerifyCodeResult verifyCodeResult) {
                    }
                });
            }
        });
        Button update = view.findViewById(R.id.update);
        update.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String email = editAccount.getText().toString();
                String verifyCode = editCode.getText().toString();
                AGConnectAuth.getInstance().getCurrentUser().updateEmail(email, verifyCode)
                    .addOnSuccessListener(new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void aVoid) {
                            Log.d(TAG, "updateEmail success");
                        }
                    }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        Log.d(TAG, "updateEmail fail" + e);
                    }
                });
                dialog.dismiss();
            }
        });
        dialog.show();
    }

}

File: Xamarin/agc-auth-xamarin/android/XamarinAuthDemo/SettingsActivity.cs
/*
       Copyright 2020-2021. Huawei Technologies Co., Ltd. All rights reserved.

       Licensed under the Apache License, Version 2.0 (the "License");
       you may not use this file except in compliance with the License.
       You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing, software
       distributed under the License is distributed on an "AS IS" BASIS,
       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
       See the License for the specific language governing permissions and
       limitations under the License.
*/
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Huawei.Agconnect.Auth;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace XamarinAuthDemo
{
    [Activity(Label = "SettingsActivity")]
    public class SettingsActivity : Activity
    {
        private const string TAG = "SettingsActivity";
        private RelativeLayout deleteUser;
        private RelativeLayout updateEmail;
        private RelativeLayout updatePhone;
        private RelativeLayout updatePassword;
        private RelativeLayout resetPassword;
        protected override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.activity_settings);
            InitView();
        }

        private void InitView()
        {
            deleteUser = FindViewById<RelativeLayout>(Resource.Id.layoutDelete);
            updateEmail = FindViewById<RelativeLayout>(Resource.Id.layoutUpdateEmail);
            updatePhone = FindViewById<RelativeLayout>(Resource.Id.layoutUpdatePhone);
            updatePassword = FindViewById<RelativeLayout>(Resource.Id.layoutUpdatePassword);
            resetPassword = FindViewById<RelativeLayout>(Resource.Id.layoutResetPassword);

            deleteUser.Click += DeleteUser_Click;
            updateEmail.Click += UpdateEmail_Click;
            updatePhone.Click += UpdatePhone_Click;
            updatePassword.Click += UpdatePassword_Click;
            resetPassword.Click += ResetPassword_Click;
        }

        private void ResetPassword_Click(object sender, EventArgs e)
        {
            View view = LayoutInflater.From(this).Inflate(Resource.Layout.dialog_reset_password, null, false);
            AlertDialog dialog = new AlertDialog.Builder(this).SetView(view).Create();
            EditText edtEmail = view.FindViewById<EditText>(Resource.Id.edtEmail);
            EditText edtNewPassword = view.FindViewById<EditText>(Resource.Id.edtNewPassword);
            EditText edtVerifyCode = view.FindViewById<EditText>(Resource.Id.edtVerifyCode);
            Button btnSend = view.FindViewById<Button>(Resource.Id.btnSend);
            Button btnUpdate = view.FindViewById<Button>(Resource.Id.btnUpdate);

            btnSend.Click += async delegate
            {
                string email = edtEmail.Text.ToString().Trim();
                // Build a verify code settings.
                VerifyCodeSettings settings = VerifyCodeSettings.NewBuilder().Action(VerifyCodeSettings.ActionResetPassword).Build();
                // Request verify code,and waiting for the verification code to be sent to your mobile phone
                try
                {
                    var requestVerifyCode = AGConnectAuth.Instance.RequestVerifyCodeAsync(email, settings);
                    VerifyCodeResult verifyCodeResult = await requestVerifyCode;
                    if (requestVerifyCode.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                    {
                        Toast.MakeText(this, "Send email verify code success! ", ToastLength.Short).Show();
                    }
                }
                catch (Exception ex)
                {
                    Toast.MakeText(this, ex.Message, ToastLength.Long).Show();
                }
            };

            btnUpdate.Click += async delegate
            {
                string email = edtEmail.Text.ToString().Trim();
                string newPassword = edtNewPassword.Text.ToString().Trim();
                string verifyCode = edtVerifyCode.Text.ToString().Trim();
                // Reset password 
                try
                {
                    var resetPasswordResult = AGConnectAuth.Instance.ResetPasswordAsync(email, newPassword, verifyCode);
                    await resetPasswordResult;
                    if (resetPasswordResult.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                    {
                        Toast.MakeText(this, "Password has been reset.", ToastLength.Short).Show();
                    }
                }
                catch (Exception ex)
                {
                    Toast.MakeText(this, "Reset password failed: " + ex.Message, ToastLength.Long).Show();
                }
                dialog.Dismiss();
            };
            dialog.Show();
        }

        private void UpdatePassword_Click(object sender, EventArgs e)
        {
            View view = LayoutInflater.From(this).Inflate(Resource.Layout.dialog_password, null, false);
            AlertDialog dialog = new AlertDialog.Builder(this).SetView(view).Create();
            EditText edtNewPassword = view.FindViewById<EditText>(Resource.Id.edtNewPassword);
            EditText edtVerifyCode = view.FindViewById<EditText>(Resource.Id.edtVerifyCode);
            Button btnSend = view.FindViewById<Button>(Resource.Id.btnSend);
            Button btnUpdate = view.FindViewById<Button>(Resource.Id.btnUpdate);
            if (AGConnectAuth.Instance.CurrentUser != null)
            {
                btnSend.Click += async delegate
                {
                    // Build a verify code settings.
                    VerifyCodeSettings settings = VerifyCodeSettings.NewBuilder().Action(VerifyCodeSettings.ActionResetPassword).Build();
                    // Request verify code,and waiting for the verification code to be sent to your mobile phone
                    try
                    {
                        var requestVerifyCode = AGConnectAuth.Instance.RequestVerifyCodeAsync(AGConnectAuth.Instance.CurrentUser.Email, settings);
                        VerifyCodeResult verifyCodeResult = await requestVerifyCode;
                        if (requestVerifyCode.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                        {
                            Toast.MakeText(this, "Send email verify code success! ", ToastLength.Short).Show();
                        }
                    }
                    catch (Exception ex)
                    {
                        Toast.MakeText(this, ex.Message, ToastLength.Long).Show();
                    }
                };

                btnUpdate.Click += async delegate
                {
                    string newPassword = edtNewPassword.Text.ToString().Trim();
                    string verifyCode = edtVerifyCode.Text.ToString().Trim();
                    // Update password
                    try
                        {
                            var updatePasswordResult = AGConnectAuth.Instance.CurrentUser.UpdatePasswordAsync(newPassword, verifyCode, AGConnectAuthCredential.EmailProvider);
                            await updatePasswordResult;
                            if (updatePasswordResult.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                            {
                                Toast.MakeText(this, "Password has been updated.", ToastLength.Short).Show();
                            }
                        }
                        catch (Exception ex)
                        {
                            Toast.MakeText(this, "Update password failed: " + ex.Message, ToastLength.Long).Show();
                        }
                    dialog.Dismiss();
                };
            }
            dialog.Show();
        }

        private void UpdatePhone_Click(object sender, EventArgs e)
        {
            View view = LayoutInflater.From(this).Inflate(Resource.Layout.dialog_phone, null, false);
            AlertDialog dialog = new AlertDialog.Builder(this).SetView(view).Create();
            EditText edtCountryCode = view.FindViewById<EditText>(Resource.Id.edtCountryCode);
            EditText edtAccount = view.FindViewById<EditText>(Resource.Id.edtAccount);
            EditText edtVerifyCode = view.FindViewById<EditText>(Resource.Id.edtVerifyCode);
            Button btnSend = view.FindViewById<Button>(Resource.Id.btnSend);
            Button btnUpdate = view.FindViewById<Button>(Resource.Id.btnUpdate);

            btnSend.Click += async delegate
            {
                string countryCode = edtCountryCode.Text.ToString().Trim();
                string phoneNumber = edtAccount.Text.ToString().Trim();

                // Build a verify code settings.
                VerifyCodeSettings settings = VerifyCodeSettings.NewBuilder().Action(VerifyCodeSettings.ActionRegisterLogin).Build();
                // Request verify code,and waiting for the verification code to be sent to your mobile phone
                try
                {
                    var requestVerifyCode = AGConnectAuth.Instance.RequestVerifyCodeAsync(countryCode, phoneNumber, settings);
                    VerifyCodeResult verifyCodeResult = await requestVerifyCode;
                    if (requestVerifyCode.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                    {
                        Toast.MakeText(this, "Send phone verify code success! ", ToastLength.Short).Show();
                    }
                }
                catch (Exception ex)
                {
                    Toast.MakeText(this, ex.Message, ToastLength.Long).Show();
                }
            };

            btnUpdate.Click += async delegate
            {
                string countryCode = edtCountryCode.Text.ToString().Trim();
                string phoneNumber = edtAccount.Text.ToString().Trim();
                string verifyCode = edtVerifyCode.Text.ToString().Trim();
                if (AGConnectAuth.Instance.CurrentUser != null)
                {
                    // Update phone
                    try
                    {
                        var requestUpdatePhone = AGConnectAuth.Instance.CurrentUser.UpdatePhoneAsync(countryCode, phoneNumber, verifyCode);
                        await requestUpdatePhone;
                        if (requestUpdatePhone.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                        {
                            Toast.MakeText(this, "Update phone success! ", ToastLength.Short).Show();
                        }
                    }
                    catch (Exception ex)
                    {
                        Toast.MakeText(this, "Update phone failed! "+ ex.Message, ToastLength.Short).Show();
                    }
                }
                dialog.Dismiss();
            };
            dialog.Show();

        }

        private void UpdateEmail_Click(object sender, EventArgs e)
        {
            View view = LayoutInflater.From(this).Inflate(Resource.Layout.dialog_email, null, false);
            AlertDialog dialog = new AlertDialog.Builder(this).SetView(view).Create();
            EditText edtAccount = view.FindViewById<EditText>(Resource.Id.edtAccount);
            EditText edtVerifyCode = view.FindViewById<EditText>(Resource.Id.edtVerifyCode);
            Button btnSend = view.FindViewById<Button>(Resource.Id.btnSend);
            Button btnUpdate = view.FindViewById<Button>(Resource.Id.btnUpdate);

            btnSend.Click += async delegate
            {
                string email = edtAccount.Text.ToString().Trim();

                // Build a verify code settings.
                VerifyCodeSettings settings = VerifyCodeSettings.NewBuilder().Action(VerifyCodeSettings.ActionRegisterLogin).Build();
                // Request verify code,and waiting for the verification code to be sent to your mobile phone
                try
                {
                    var requestVerifyCode = AGConnectAuth.Instance.RequestVerifyCodeAsync(email, settings);
                    VerifyCodeResult verifyCodeResult = await requestVerifyCode;
                    if (requestVerifyCode.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                    {
                        Toast.MakeText(this, "Send email verify code success! ", ToastLength.Short).Show();
                    }
                }
                catch (Exception ex)
                {
                    Toast.MakeText(this, ex.Message, ToastLength.Long).Show();
                }
            };

            btnUpdate.Click += async delegate
            {
                string email = edtAccount.Text.ToString().Trim();
                string verifyCode = edtVerifyCode.Text.ToString().Trim();
                if (AGConnectAuth.Instance.CurrentUser != null)
                {
                    // Update phone
                    try
                    {
                        var requestUpdateEmail = AGConnectAuth.Instance.CurrentUser.UpdateEmailAsync(email, verifyCode);
                        await requestUpdateEmail;
                        if (requestUpdateEmail.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                        {
                            Toast.MakeText(this, "Update email success! ", ToastLength.Short).Show();
                        }
                    }
                    catch (Exception ex)
                    {
                        Toast.MakeText(this, "Update email failed! " + ex.Message, ToastLength.Short).Show();
                    }
                }
                dialog.Dismiss();
            };
            dialog.Show();
        }

        private async void DeleteUser_Click(object sender, EventArgs e)
        {
            // Delete user
            try
            {
                var deleteUserRequest = AGConnectAuth.Instance.DeleteUserAsync();
                await deleteUserRequest;
                if (deleteUserRequest.Status.Equals(System.Threading.Tasks.TaskStatus.RanToCompletion))
                {
                    Toast.MakeText(this, "Delete user success! ", ToastLength.Short).Show();
                    StartActivity(new Intent(this, typeof(LoginActivity)));
                    Finish();
                }
            }
            catch (Exception ex)
            {
                Toast.MakeText(this, "Delete user failed! " + ex.Message,ToastLength.Long).Show();
            }
        }
    }
}